src/common/libflux/sync.c

```diff
@@ -0,0 +1,143 @@
+/************************************************************\
+ * Copyright 2021 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* sync.c - future synchronized with heartbeat
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <flux/core.h>
+
+struct flux_sync {
+    flux_t *h;
+    uint32_t seq;
+    int count;
+    double last;
+    double minimum;
+};
+
+static const char *auxkey = "flux::sync";
+
+static void sync_destroy (struct flux_sync *sync)
+{
+    if (sync) {
+        int saved_errno = errno;
+        (void)flux_event_unsubscribe (sync->h, "heartbeat.pulse");
+        free (sync);
+        errno = saved_errno;
+    }
+}
+
+static struct flux_sync *sync_create (flux_t *h, double minimum)
+{
+    struct flux_sync *sync;
+
+    if (!(sync = calloc (1, sizeof (*sync))))
+        return NULL;
+    sync->h = h;
+    sync->minimum = minimum;
+    if (flux_event_subscribe (h, "heartbeat.pulse") < 0)
+        goto error;
+    return sync;
+error:
+    sync_destroy (sync);
+    return NULL;
+}
+
+/* Special note about the non-reactive case:  events are delivered to all
+ * matching message handlers, not first match like requests.  Therefore, the
+ * future implementation must requeue events even if they were matched in the
+ * temporary reactor, in case another matching handler exists in the main
+ * reactor.  Thus, calling flux_future_get() in a loop on the sync object,
+ * where the main reactor's dispatcher doesn't retire the event, would fulfill
+ * the future using the same message over and over unless we take steps to
+ * prevent that by watching the event sequence number.
+ */
+static void heartbeat_cb (flux_t *h,
+                          flux_msg_handler_t *mh,
+                          const flux_msg_t *msg,
+                          void *arg)
+{
+    double now = flux_reactor_now (flux_get_reactor (h));
+    flux_future_t *f = arg;
+    struct flux_sync *sync = flux_future_aux_get (f, auxkey);
+    uint32_t seq;
+
+    if (flux_msg_authorize (msg, FLUX_USERID_UNKNOWN) < 0
+            || flux_event_decode (msg, NULL, NULL) < 0
+            || flux_msg_get_seq (msg, &seq) < 0
+            || (sync->count > 0 && seq <= sync->seq)
+            || (sync->count > 0 && sync->minimum > 0.
+                                && now - sync->last < sync->minimum)) {
```

Comment:
Another extra 4 space indent here.. though maybe you are doing that on purpose... ignore if so.

Reply:
It was kind of on purpose but I don't know why.   Will fix.